File: src/niepce/ui/dialogs/editlabels.cpp
/*
 * niepce - niepce/ui/dialogs/editlabels.cpp
 *
 * Copyright (C) 2009-2023 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <algorithm>

#include <boost/format.hpp>

#include <glibmm/i18n.h>
#include <gtkmm/entry.h>
#include <gtkmm/label.h>

#include "fwk/base/debug.hpp"
#include "fwk/toolkit/application.hpp"
#include "fwk/toolkit/gdkutils.hpp"
#include "editlabels.hpp"

using eng::LibraryClientPtr;

namespace ui {

EditLabels::EditLabels(const eng::LibraryClientHost& libclient)
    : fwk::Dialog("/net/figuiere/Niepce/ui/editlabels.ui", "editLabels")
    , m_lib_client(libclient)
{
    auto& provider = libclient.getDataProvider();
    auto count = provider.label_count();
    for (size_t i = 0; i < count; i++) {
        m_labels.push_back(rust::Box<eng::Label>::from_raw(provider.label_at(i)));
    }

    std::fill(m_status.begin(), m_status.end(), false);
}

void EditLabels::run_modal(GtkWindow* parent, rust::Fn<void(EditLabelsPtr, int32_t)> on_ok, EditLabelsPtr self) const
{
    run_modal_(parent, [self, on_ok] (int32_t r) { on_ok(self, r); });
}

void EditLabels::setup_widget()
{
    Glib::RefPtr<Gtk::Builder> _builder = builder();
    DBG_OUT("setup Edit Labels dialog");
    add_header(_("Edit Labels"));

    const char * colour_fmt = "colorbutton%1%";
    const char * value_fmt = "value%1%";
    for (size_t i = 0; i < NUM_LABELS; i++) {
        bool has_label = m_labels.size() > i;

        Gtk::ColorButton *colourbutton;
        Gtk::Entry *labelentry;

        m_colours[i] = colourbutton = _builder->get_widget<Gtk::ColorButton>(str(boost::format(colour_fmt) % (i+1)));
        DBG_ASSERT(labelentry, "couldn't find label");
        m_entries[i] = labelentry = _builder->get_widget<Gtk::Entry>(str(boost::format(value_fmt) % (i+1)));

        if(has_label) {
            Gdk::RGBA colour = fwk::rgbcolour_to_gdkcolor(m_labels[i]->colour());
            colourbutton->set_rgba(colour);
            labelentry->set_text(std::string(m_labels[i]->label()));
        }
        colourbutton->signal_color_set().connect(
            sigc::bind(sigc::mem_fun(*this, &EditLabels::label_colour_changed), i));
        labelentry->signal_changed().connect(
            sigc::bind(sigc::mem_fun(*this, &EditLabels::label_name_changed), i));
    }
    DBG_OUT("all colours setup");
    gtkDialog().signal_response().connect(sigc::mem_fun(*this, &EditLabels::update_labels));
}


void EditLabels::label_name_changed(size_t idx)
{
    m_status[idx] = true;
}


void EditLabels::label_colour_changed(size_t idx)
{
    m_status[idx] = true;
}

void EditLabels::update_labels(int /*response*/)
{
    rust::Box<fwk::UndoTransaction> undo = fwk::UndoTransaction_new(_("Change Labels"));
    for(size_t i = 0; i < 5; i++) {
        if(m_status[i]) {
            bool has_label = m_labels.size() > i;

            DBG_OUT("updating label %lu", i);
            std::string new_name = m_entries[i]->get_text();
            // a newname is NOT valid.
            if(new_name.empty()) {
                continue;
            }
            std::string new_colour(fwk::gdkcolor_to_rgbcolour(m_colours[i]->get_rgba())->to_string());

            auto client = &m_lib_client.client();
            if(has_label) {
                std::string current_name(m_labels[i]->label());
                std::string current_colour(m_labels[i]->colour().to_string());
                auto label_id = m_labels[i]->id();

                auto command = fwk::UndoCommand_new(
                    std::make_unique<fwk::RedoFnVoid>(
                        [client, new_name, new_colour, label_id] () {
                            client->update_label(label_id, new_name, new_colour);
                        }),
                    std::make_unique<fwk::UndoFnVoid>(
                        [client, current_name, current_colour, label_id] () {
                            client->update_label(label_id, current_name, current_colour);
                        })
                    );
                undo->add(std::move(command));
            } else {
                auto command = fwk::UndoCommand_new_int(
                    std::make_unique<fwk::RedoFnInt>(
                        [client, new_name, new_colour] () {
                            return client->create_label_sync(new_name, new_colour);
                        }),
                    std::make_unique<fwk::UndoFnInt>(
                        [client] (int64_t label) {
                            client->delete_label(label);
                        })
                    );
                undo->add(std::move(command));
            }
        }
    }
    if (!undo->is_empty()) {
        undo->execute();
        fwk::Application::app()->begin_undo(std::move(undo));
    }
}

/*
  Local Variables:
  mode:c++
  c-file-style:"stroustrup"
  c-file-offsets:((innamespace . 0))
  indent-tabs-mode:nil
  fill-column:80
  End:
*/
}

File: niepce-main/src/niepce/ui/dialogs/edit_labels.rs
/*
 * niepce - niepce/ui/dialogs/edit_labels.rs
 *
 * Copyright (C) 2024-2025 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::cell::RefCell;
use std::rc::Rc;
use std::sync::{Arc, Weak};

use adw::prelude::*;
use gettextrs::gettext as i18n;
use npc_fwk::{adw, glib, gtk4};

use npc_engine::catalog;
use npc_engine::libraryclient::{
    ClientInterface, ClientInterfaceSync, LibraryClient, LibraryClientHost,
};
use npc_fwk::base::RgbColour;
use npc_fwk::toolkit::{
    AppController, Controller, ControllerImplCell, DialogController, Storage, UiController,
    UndoCommand, UndoTransaction,
};
use npc_fwk::{controller_imp_imp, send_async_local};

use crate::NiepceApplication;

const NUM_LABELS: usize = 5;

pub enum InMsg {
    ColourChanged(usize),
    NameChanged(usize),
    CloseRequest,
}

pub struct EditLabels {
    imp_: ControllerImplCell<InMsg, ()>,
    client: Arc<LibraryClient>,
    app: Weak<NiepceApplication>,
    labels: Vec<catalog::Label>,
    colours: Vec<gtk4::ColorDialogButton>,
    entries: Vec<gtk4::Entry>,
    status: RefCell<[bool; NUM_LABELS]>,
    dialog: adw::Window,
}

impl Controller for EditLabels {
    type InMsg = InMsg;
    type OutMsg = ();

    controller_imp_imp!(imp_);

    fn dispatch(&self, msg: InMsg) {
        match msg {
            InMsg::ColourChanged(idx) | InMsg::NameChanged(idx) => self.changed(idx),
            InMsg::CloseRequest => {
                self.update_labels();
                self.close();
            }
        }
    }
}

impl UiController for EditLabels {
    fn widget(&self) -> &gtk4::Widget {
        self.dialog.upcast_ref()
    }
}

impl DialogController for EditLabels {
    fn dialog(&self) -> &adw::Window {
        &self.dialog
    }
}

impl EditLabels {
    pub fn new(client: &Rc<LibraryClientHost>, app: Weak<NiepceApplication>) -> Rc<EditLabels> {
        let builder = gtk4::Builder::from_resource("/net/figuiere/Niepce/ui/editlabels.ui");
        let provider = client.ui_provider();
        let mut labels = vec![];

        assert!(provider.label_count() >= NUM_LABELS);

        for idx in 0..NUM_LABELS {
            labels.push(provider.label_at(idx));
        }
        get_widget!(builder, adw::Window, edit_labels);

        let mut ctrl = EditLabels {
            imp_: ControllerImplCell::default(),
            client: client.client().clone(),
            app,
            labels,
            entries: vec![],
            colours: vec![],
            status: RefCell::new([false; 5]),
            dialog: edit_labels,
        };

        ctrl.build_widget(builder);

        let ctrl = Rc::new(ctrl);
        <Self as DialogController>::start(&ctrl);

        ctrl
    }

    fn build_widget(&mut self, builder: gtk4::Builder) {
        let colour_dialog = gtk4::ColorDialog::new();
        for idx in 0..NUM_LABELS {
            self.colours.push(
                builder
                    .object::<gtk4::ColorDialogButton>(format!("colorbutton{}", idx + 1))
                    .unwrap(),
            );
            self.entries.push(
                builder
                    .object::<gtk4::Entry>(format!("value{}", idx + 1))
                    .unwrap(),
            );

            let colour = self.labels[idx].colour();
            self.colours[idx].set_dialog(&colour_dialog);
            self.colours[idx].set_rgba(&(colour.clone()).into());
            self.entries[idx].set_text(self.labels[idx].label());

            let sender = self.sender();
            self.colours[idx].connect_notify(Some("rgba"), move |_, _| {
                send_async_local!(InMsg::ColourChanged(idx), sender);
            });
            let sender = self.sender();
            self.entries[idx].connect_changed(move |_| {
                send_async_local!(InMsg::NameChanged(idx), sender);
            });
        }

        let sender = self.sender();
        self.dialog.connect_close_request(move |_| {
            send_async_local!(InMsg::CloseRequest, sender);
            glib::Propagation::Proceed
        });
    }

    fn changed(&self, idx: usize) {
        self.status.borrow_mut()[idx] = true;
    }

    fn update_labels(&self) {
        let mut undo = UndoTransaction::new(&i18n("Change Labels"));
        let statuses = self.status.borrow();
        for status in statuses.iter().enumerate() {
            if !status.1 {
                continue;
            }
            let update = true;
            let new_name = self.entries[status.0].text().to_string();
            let new_colour: RgbColour = self.colours[status.0].rgba().into();
            let current_name = self.labels[status.0].label().to_string();
            let current_colour: RgbColour = self.colours[status.0].rgba().into();
            let label_id = self.labels[status.0].id();

            let client_undo = self.client.clone();
            let client_redo = self.client.clone();

            let command = if update {
                UndoCommand::new(
                    Box::new(move || {
                        client_redo.update_label(label_id, new_name.clone(), new_colour.clone());
                        Storage::Void
                    }),
                    Box::new(move |_| {
                        client_undo.update_label(
                            label_id,
                            current_name.clone(),
                            current_colour.clone(),
                        );
                    }),
                )
            } else {
                UndoCommand::new(
                    Box::new(move || {
                        client_redo
                            .create_label_sync(new_name.clone(), new_colour.clone())
                            .into()
                    }),
                    Box::new(move |label| {
                        client_undo.delete_label(label.into());
                    }),
                )
            };
            undo.add(command);
        }
        if !undo.is_empty() {
            undo.execute();
            if let Some(app) = Weak::upgrade(&self.app) {
                app.begin_undo(undo);
            }
        }
    }
}
